python/ray/data/_internal/planner/exchange/shuffle_task_spec.py

```diff
@@ -33,22 +42,37 @@ def map(
         idx: int,
         block: Block,
         output_num_blocks: int,
+        target_max_block_size: int,
         upstream_map_fn: Optional[Callable[[Iterable[Block]], Iterable[Block]]],
         random_shuffle: bool,
         random_seed: Optional[int],
     ) -> List[Union[BlockMetadata, Block]]:
-        # TODO: Support fusion with other upstream operators.
         stats = BlockExecStats.builder()
         if upstream_map_fn:
-            mapped_blocks = list(upstream_map_fn([block]))
-            if len(mapped_blocks) > 1:
-                builder = BlockAccessor.for_block(mapped_blocks[0]).builder()
-                for b in mapped_blocks:
-                    builder.add_block(b)
-                block = builder.build()
-            else:
-                block = mapped_blocks[0]
+            # TODO: Support dynamic block splitting in
+            # all-to-all ops, to avoid having to re-fuse
+            # upstream blocks together.
+            upstream_map_iter = upstream_map_fn([block])
+            mapped_block = next(upstream_map_iter)
+            builder = BlockAccessor.for_block(mapped_block).builder()
+            builder.add_block(mapped_block)
+            for mapped_block in upstream_map_iter:
+                builder.add_block(mapped_block)
+            # Drop the upstream inputs to reduce memory usage.
+            del mapped_block
+            block = builder.build()
         block = BlockAccessor.for_block(block)
+        if block.size_bytes() > target_max_block_size:
+            logger.get_logger().warn(
```

Comment:
I am worried about if this warning would confuse users. Can we log it as `debug`?

Reply:
I think print a warning makes sense. But we may want to make it less verbose by, for example, checking size_bytes >  2 * target_max_block_size. 
Also, suggesting increasing parallelism seems to make more sense than disabling fusion.

Reply:
I'll change to 2 * target_max_block_size.

But materialize() is the best solution here (see other comment). Hard to recommend an exact fix via increasing parallelism.